perfmetrics/scripts/load_tests/python/load_generator/load_generator.py

```diff
@@ -0,0 +1,352 @@
+      for.
+    num_tasks_per_thread: An integer defining the number of times the given
+      task to be run inside each thread of each process during load test.
+  """
+
+  def __init__(self,
+               num_processes,
+               num_threads_per_process,
+               run_time=sys.maxsize,
+               num_tasks_per_thread=sys.maxsize):
+    self.num_processes = num_processes
+    self.num_threads_per_process = num_threads_per_process
+    self.run_time = min(sys.maxsize, run_time)
+    self.num_tasks_per_thread = min(sys.maxsize, num_tasks_per_thread)
+
+    if (self.run_time == sys.maxsize) & \
+        (self.num_tasks_per_thread == sys.maxsize):
+      raise ValueError('Out of run_time and num_tasks_per_thread '
+                       'arguments, one has to be passed.')
+
+    self.total_num_tasks = sys.maxsize
+    if self.num_tasks_per_thread != sys.maxsize:
+      self.total_num_tasks = self.num_tasks_per_thread * \
+                             self.num_threads_per_process * self.num_processes
+
+  def pre_load_test(self):
+    """Task to perform before running load test.
+    """
+    pass
+
+  def generate_load(self, task):
+    """Generates load with the given task.
+
+    The load is generated on CPU and other resources (depending upon the task
+    used) by running process(s) and thread(s) where task runs inside thread(s)
+    and thread(s) runs inside process(s).
+
+    Args:
+      task: Implementation of task.LoadTestTask.
+
+    Returns:
+      Returns start_time, end_time of load test, latencies and results of all
+        the tasks performed over the span of load test.
+
+    Raises:
+      RuntimeError: If the given task is not completed even once during the
+        course of load test.
+    """
+    tasks_results_queue = multiprocessing.Manager().Queue()
+    pre_tasks_results_queue = multiprocessing.Manager().Queue()
+    post_tasks_results_queue = multiprocessing.Manager().Queue()
+
+    processes = []
+    for process_id in range(self.num_processes):
+      process = multiprocessing.Process(
+          target=LoadGenerator._process_task,
+          args=(task, process_id, self.num_threads_per_process,
+                self.num_tasks_per_thread, pre_tasks_results_queue,
+                tasks_results_queue, post_tasks_results_queue))
+      processes.append(process)
+
+    # Checkpoints initialisation. Note: Only used if self.run_time is set.
+    log_loading = self.run_time != sys.maxsize
+    loading_checkpoints = list(
+        map(lambda t: (t * self.run_time), lg_const.TIME_LOADING_PERCENTAGES))
+    curr_loading_idx = 0
+
+    for process in processes:
+      process.start()
+    logging.debug('%s number of processes started for task %s', len(processes),
+                  task.task_name)
+
+    start_time = curr_time = time.time()
+    loading_checkpoints = [t + start_time for t in loading_checkpoints]
+    while ((curr_time - start_time) < self.run_time) & \
```

Comment:
can you elaborate what is happeninng here.

Reply:
Added comment:
Loop till the condition of termination of load test is not met. The
condition is either the load test has run for self.run_time or completed
the total number of tasks assigned.